### ISSUE TITLE
Import module .from=`variable`

### ISSUE THREAD

**Issue by clamus** [rticulate/import#13]
I am trying to import a function for a module, where the the module path is in a variable. This does not work, as the variable is interpreted as a package instead of a string pointing to the module's file. In particular, doing this does not work:
path_to_module <- "~/repos/R/my_repo/my_module.R"
import::from(.from=path_to_module, my_function, .into="")

Is there an easy way to go around this restriction?

**Comment by smbache** [comment]
Currently `import` behaves similarly to `library` using non-standard evaluation to use the expression rather than the value. E.g. 

```R
foo <- "bar"
library(foo)
```
Will attach `foo` and not `bar`. Generally, this is ok since these statements should express what they are importing clearly. 

That said, I guess there could perhaps be use cases where it could be acceptable; and I guess semantics like e.g. `dplyr`'s `verb_` family of functions that use standard evaluation could be an option. 

Do you have examples where it's ok to mask what you import from in a variable?

**Comment by clamus** [comment]
Thanks for the quick reply. The use case I am thinking about is to be able to make relative import, like one can do in python. Suppose we have a module organized like this:
> somewhere/repos/my_module/io.R
> somewhere/repos/my_module/utils.R
> somewhere/repos/my_module/other.R

Then, in `io.R`, one would like to use func_util from `utils.R`. Doing `import::here(.from="somewhere/repos/my_module/utils.R", func_util)` in `io.R` does the job. However, and very importantly, suppose you want to share your module with a college. Then, the code would not work, unless the college copies the module in an identical hierarchy. This is when a relative path or a setting the path in variables in some system/user dependent way could be very useful. And probably, setting relative paths my be easier, and would look very similar to python.

**Comment by smbache** [comment]
A second option would be, then, to let the `.library` be used for this when importing from a "module", e.g.

```R
lib <- "/path/to/modules"

import::from(foo.R, bar, baz, .library = lib)
```

This would make good sense as analogue to using a specific library for packages.

**Comment by clamus** [comment]
Thanks @smbache! I tried what you suggested but it did not work. This is what I tried to do:

I have the following R files in a module:

> /path/to/module/file1.R
> /path/to/module/file2.R

In `file1.R` I have a function `func1`. I would like to import `func1` in `file2.R`. To do this, I have the following in `file2.R`:

`module_dir <- dirname(sys.frame(1)$ofile)`
`import::from(file1.R, func1, .library=module_dir)`

While the `module_dir` is correct in that is shows `/path/to/module`, I still get the error:

`Error: there is no package called ‘file1.R’`

What do you think I am doing incorrectly?

**Comment by smbache** [comment]
It's not working because it is not implemented yet. It was a suggestion for the future :)

**Comment by clamus** [comment]
Hehehehehe! Probably the best answer I have heard in a while!!! 

Now that you mention it is not implemented, it might be good to have something that allows imports from relative paths. This might be a really clear way to import something like we can do in python. Like:

`import:here(func1, .from="../submodule2/file_a.R")`

This would allow people to make python like module really easily, and would also make it really easy to share this modules.

**Comment by J-Moravec** [comment]
Another problematic case is that it depends from where are files ran.

Imagine project organisation:
```
src/lib/utils.r
src/script.r
data/...
results/...
```
Now, if the script has:
```
import::from("src/lib/utils.r", function)
```
It all works great if and only if the script `src/script.r` is run in the main folder (i.e., with `Rscript src/script.r"). If however, one runs script somewhere else (`src/script.r/results`), it fails because it cannot find the libs. This can be generally avoided. However, in my case, I am using submakefile to build stuff for two similar, but not same, datasets.

A similar outcome can be gained with `source` and `sys.source` by specifying `chdir=TRUE`. While this might have unintended consequences, if import is used to import function that have no side-effects, it should be fine.

I have tried to cook something similar by looking at path from `commandArgs(FALSE)`, but since variables cannot be used on import path, this won't work.

**Comment by quartin** [comment]
> A second option would be, then, to let the `.library` be used for this when importing from a "module", e.g.
> 
> ```r
> lib <- "/path/to/modules"
> 
> import::from(foo.R, bar, baz, .library = lib)
> ```
> 
> This would make good sense as analogue to using a specific library for packages.

Any plans to implement this? :)

**Comment by torfason** [comment]
It seems like pretty much all the use cases referred in this issue can be resolved using the new `.character_only` parameter to `import::from()`, which is now available in the development version. 

(The only issue that is *not* resolved, is the one referenced by @J-Moravec, which seems like a duplicate of #15).

Example of how to address the other use cases (as I understand them) follows, with some setup code to create the directory and file in order to make this reproducible:

``` r
dynamicDir = "arbitraryLocation"
dynamicFile = file.path(dynamicDir,"arbitraryFile.R")
dir.create(dynamicDir, showWarnings=FALSE)
cat("myfunc <- function(x) { x + 1 }", file=dynamicFile)
import::from(dynamicFile, "myfunc", .character_only=TRUE)
myfunc(3)
#> [1] 4
```

<sup>Created on 2020-06-25 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

**Comment by J-Moravec** [comment]
@torfason should I restate my problem there and create MRE (I think my problem is slightly different)?

**Comment by torfason** [comment]
@J-Moravec yes that sounds like a good plan, either here or there depending on whether the problem is (a) the directory where the import statement is initiated or [which I think might belong here] or (b) the directory where the contents of the module are evaluated [which I think might belong in #15].

I do wonder in reading your issue again, whether this is something that could be solved by using the `here` package to set a root directory for your project and then calling: 
`import::from(here::here("src/lib/utils.r")), "function", .character_only=TRUE)`

`here::here()` returns an absolute path and does not (much) care what the current working directory is, it will determine the root directory for the project pretty intelligently in my experience.

**Comment by J-Moravec** [comment]
For MRE see:

https://github.com/J-Moravec/RImportUserCase

If I am not mistaken, when `import::from(path, function)` parse the script in `path` to find `function`, it does it relative to the `dirname(path)`, so that no chained imports are broken (although I would swear it wasn't like that? Because that is exactly what `chdir=TRUE` in `sys.source` does. In the MRE, if this behaviour wasn't the case, importing the `to_title` file should fail)

The idea here is to somehow tell `import` to take the `path` relative to the file's path and not the current working directory, since when running with `Rscript`, the working directory might change which would break the imports. See the `src/run_me_fixed.r` which, at least with respect to `Rscript`, change the workdir to the dirname of script and thus can be started from anywhere while not breaking the imports.

**Comment by torfason** [comment]
Looks good, I think I should add this to the unit tests.

Yes, the new `.chdir` parameter was intended in part to address this issue, so it is great if it does.

I'm actually thinking that it might be nice to allow character args to `.chdir`, so boolean args would work as currently, but `.chdir="/modules"` would source files relative to the `/modules` directory. Any thoughts on that?

**Comment by torfason** [comment]
@J-Moravec, I was trying to add the MRE into the tests, but on my setup the MRE this fails regardless of any directory issues. Even if all the scripts are in the same directory, the `to_title()` function is not found. It seems to me that recursive `import::from()` statements simply do not work, regardless of any directory issues. Are you actually seeing this working when you keep everything in the same directory?

**Comment by J-Moravec** [comment]
I can confirm that the example works with R 3.4 and R 4.0. See my session info (import loaded just to report its version):

```
R version 4.0.2 (2020-06-22)
Platform: x86_64-pc-linux-gnu (64-bit)
Running under: Ubuntu 18.04.5 LTS

Matrix products: default
BLAS:   /usr/lib/x86_64-linux-gnu/openblas/libblas.so.3
LAPACK: /usr/lib/x86_64-linux-gnu/libopenblasp-r0.2.20.so

locale:
 [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
 [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    
 [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   
 [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 
 [9] LC_ADDRESS=C               LC_TELEPHONE=C            
[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] import_1.1.0

loaded via a namespace (and not attached):
[1] compiler_4.0.2
```

**Comment by torfason** [comment]
I've looked more into this, and found that recursive importing (importing from file `a.R`, where `a.R` contains an import from another file `b.R`  seems to have stopped working on commit (01c27bf). This is the same commit that @mmuurr mentioned in #28. 

In my test, the issue was *not* related to any directory issues, it happens even if all the files are together in the current working directory. Unfortunately, this also means that the issue is *not* fixed with my recent pull request, and I'm not sure how exactly to fix it. My guess however, is that a fix for the same-directory case would also fix the different-directory case.

**Comment by smbache** [comment]
I suspect this is because if

1) import things from script.R
2) script.R imports from `foo` and attaches it
3) import from (1) finishes and cleans up by removing things (see [here](https://github.com/smbache/import/pull/26/files#diff-50b9fb6815db35ecc7fbc8d3aa16c614R167))) and since it doesn't know about `foo`, it is detached 
4) some function in script.R is called and expects to be able to find things from `foo` and it will break.

If hypothesis is correct, one approach could be to place a `.module` (similar to the `.packageName`) in the script env, and have `import::from` prefix the search path entries eg `.module:import-name` , and never remove these things in the cleanup linked above.

WDYT @torfason

**Comment by J-Moravec** [comment]
@smbache This sounds to me like the same reason why the `library()` doesn't work in `script.R`?

**Comment by smbache** [comment]
@J-Moravec Yes that's very likely!

**Comment by torfason** [comment]
@smbache This sounds like a very reasonable explanation. It fits with the fact that when testing the MRE, I saw that in version before 01c27bf, the `to_title()` function (the helper function imported recursively) was listed in the object explorer under the `import` environment, but in 01c27bf and later, `to_title()`. But I'm not sure what a 

I fiddled around, seeing what would happen if the recursive import was into another environment (changing `title_text.r` to import using `import::from("to_title.r", to_title, .into="other_environment")`. But that leads to other errors:

`> packageVersion("import")
[1] ‘1.1.9000’
> import::from("text.r", print_text)
> import::from("title_text.r", print_title_text)
 Error in as.environment(where) : 
  no item called "other_environment" on the search list`

That seems like another issue, that would not be fixed by a fix focused on allowing recursively imported functions to live, and maintaining a stack of environments to use for different levels in the recursion seems complex and error prone.

(after running `import::from("to_title.r", to_title, .into="other_environment")` directly, not inside a recursive statement, the above example works, because the other environment already exists. But it's not very reassuring.)

My lean is to mark that an import is going on, and _throw an error on a recursive import_. If this were not a regression from the released version, that would be my strong lean. The prior functionality, where an unknown function ( `to_title()`) appeared in the `import` environment after an import, seems like an undocumented side effect. I worry that implementing this robustly will become both complex and error prone. Is this within the scope of `import`? With more user friendly package writing tools, shouldn't people with complex modules simply write packages?

All that said, R environments and scope are something that I am not an expert in. Perhaps there is a neat way to do this - if so, I'd be happy to help out by incorporating tests to root out pesky edge cases.

**Comment by torfason** [comment]
I can confirm that when using `here()` instead of `from()` in the module file, recursive imports *do* work. Made a small change to the MRE, that is now in the unit tests (available in `torfason/import/dev`), which makes the MRE work. The relevant change is in the `title_text_here.R` file, which looks like this:

```
import::here(.from=to_title.R, to_title)

print_title_text = function(text){
  print(to_title(text))
}
```

**Comment by torfason** [comment]
#34 has now been closed as fixed, and all problems relating to this issue have been addressed in the new release, 1.2.0, which is now on CRAN.
